source/fab/tools/compiler.py

```diff
@@ -92,68 +93,95 @@ def check_available(self) -> bool:
-            a different error happened when trying to get the compiler version.
+        :returns: a tuple of at least 2 integers, representing the version
+            e.g. (6, 10, 1) for version '6.10.1'.
 
-        :raises RuntimeError: if the compiler was not found.
+        :raises RuntimeError: if the compiler was not found, or if it returned
+            an unrecognised output from the version command.
         """
-        if self._version:
+        if self._version is not None:
             return self._version
 
-        try:
-            res = self.run("--version", capture_output=True)
-        except FileNotFoundError as err:
-            raise RuntimeError(f'Compiler not found: {self.name}') from err
-        except RuntimeError as err:
-            self.logger.warning(f"Error asking for version of compiler "
-                                f"'{self.name}': {err}")
-            return ''
+        # Run the compiler to get the version and parse the output
+        # The implementations depend on vendor
+        output = self.run_version_command()
+        version_string = self.parse_version_output(self.category, output)
 
-        # Pull the version string from the command output.
-        # All the versions of gfortran and ifort we've tried follow the
-        # same pattern, it's after a ")".
+        # Expect the version to be dot-separated integers.
+        # todo: Not all will be integers? but perhaps major and minor?
         try:
-            version = res.split(')')[1].split()[0]
-        except IndexError:
-            self.logger.warning(f"Unexpected version response from "
-                                f"compiler '{self.name}': {res}")
-            return ''
-
-        # expect major.minor[.patch, ...]
-        # validate - this may be overkill
-        split = version.split('.')
-        if len(split) < 2:
-            self.logger.warning(f"unhandled compiler version format for "
-                                f"compiler '{self.name}' is not "
-                                f"<n.n[.n, ...]>: {version}")
-            return ''
-
-        # todo: do we care if the parts are integers? Not all will be,
-        # but perhaps major and minor?
-
-        self.logger.info(f'Found compiler version for {self.name} = {version}')
+            version = tuple(int(x) for x in version_string.split('.'))
+        except ValueError as err:
+            raise RuntimeError(f"Unexpected version output format for "
+                               f"compiler '{self.name}'. Should be numeric "
+                               f"<n.n[.n, ...]>: {version_string}") from err
+
+        # Expect at least 2 integer components, i.e. major.minor[.patch, ...]
+        if len(version) < 2:
+            raise RuntimeError(f"Unexpected version output format for "
+                               f"compiler '{self.name}'. Should have at least "
+                               f"two parts, <n.n[.n, ...]>: {version_string}")
+
```

Comment:
Why is this changed from a log warning to a runtime error?

Reply:
I didn't mention in the summary, but part of this change is the design decision that `get_version()` should only return a valid version tuple. if it can't determine a version, it raises a runtime error.
Previously it could return an empty string in some cases, but we thought that just made things unclear, and increased the amount of checking you might need to do when calling `get_version()`.

Reply:
Agreed - if we can't detect the version number (be it because the compiler has changed, or we have a bug in Fab), we need to abort. In LFRic we will have compiler flags added depending on compiler version, so it is important that we get the version correct.